File: src/tools/integrations.ts
import type { Client } from "../client.js";

export type GoogleDriveStatusResult = Client extends { googleDriveStatus(): Promise<infer R> } ? R : never;
export type GoogleDriveConnectResult = Client extends { googleDriveConnect(): Promise<infer R> } ? R : never;
export type GoogleGmailStatusResult = Client extends { googleGmailStatus(): Promise<infer R> } ? R : never;
export type GoogleGmailConnectResult = Client extends { googleGmailConnect(): Promise<infer R> } ? R : never;
export type MicrosoftOutlookStatusResult = Client extends { microsoftOutlookStatus(): Promise<infer R> } ? R : never;
export type MicrosoftOutlookConnectResult = Client extends { microsoftOutlookConnect(): Promise<infer R> } ? R : never;

export function newErrorResult(msg: string) {
  return {
    content: [{ type: "text" as const, text: `Error: ${msg}` }],
    isError: true,
  };
}

export async function googleDriveStatus(client: Client) {
  try {
    const result = await client.googleDriveStatus();
    return { content: [{ type: "json" as const, text: JSON.stringify(result) }] };
  } catch (e) {
    return newErrorResult(e instanceof Error ? e.message : String(e));
  }
}

export async function googleDriveConnect(client: Client) {
  try {
    const result = await client.googleDriveConnect();
    return { content: [{ type: "json" as const, text: JSON.stringify(result) }] };
  } catch (e) {
    return newErrorResult(e instanceof Error ? e.message : String(e));
  }
}

export async function googleGmailStatus(client: Client) {
  try {
    const result = await client.googleGmailStatus();
    return { content: [{ type: "json" as const, text: JSON.stringify(result) }] };
  } catch (e) {
    return newErrorResult(e instanceof Error ? e.message : String(e));
  }
}

export async function googleGmailConnect(client: Client) {
  try {
    const result = await client.googleGmailConnect();
    return { content: [{ type: "json" as const, text: JSON.stringify(result) }] };
  } catch (e) {
    return newErrorResult(e instanceof Error ? e.message : String(e));
  }
}

export async function microsoftOutlookStatus(client: Client) {
  try {
    const result = await client.microsoftOutlookStatus();
    return { content: [{ type: "json" as const, text: JSON.stringify(result) }] };
  } catch (e) {
    return newErrorResult(e instanceof Error ? e.message : String(e));
  }
}

export async function microsoftOutlookConnect(client: Client) {
  try {
    const result = await client.microsoftOutlookConnect();
    return { content: [{ type: "json" as const, text: JSON.stringify(result) }] };
  } catch (e) {
    return newErrorResult(e instanceof Error ? e.message : String(e));
  }
}

File: internal/tools/integrations.go
package tools

import (
	"context"

	"github.com/modelcontextprotocol/go-sdk/mcp"

	"github.com/zamedic/labradoc-mcp/internal/labradoc"
)

// GoogleDriveStatusResult represents output for google_drive_status tool.
type GoogleDriveStatusResult = labradoc.IntegrationStatus

// GoogleDriveConnectResult represents output for google_drive_connect tool.
type GoogleDriveConnectResult = labradoc.BillingCheckoutResponse

// GoogleGmailStatusResult represents output for google_gmail_status tool.
type GoogleGmailStatusResult = labradoc.IntegrationStatus

// GoogleGmailConnectResult represents output for google_gmail_connect tool.
type GoogleGmailConnectResult = labradoc.BillingCheckoutResponse

// MicrosoftOutlookStatusResult represents output for microsoft_outlook_status tool.
type MicrosoftOutlookStatusResult = labradoc.IntegrationStatus

// MicrosoftOutlookConnectResult represents output for microsoft_outlook_connect tool.
type MicrosoftOutlookConnectResult = labradoc.BillingCheckoutResponse

// GoogleDriveStatus checks if Google Drive is connected.
func GoogleDriveStatus(ctx context.Context, client *labradoc.Client) (*mcp.CallToolResult, GoogleDriveStatusResult, error) {
	result, err := client.GoogleDriveStatus(ctx)
	if err != nil {
		return errorResult(err), labradoc.IntegrationStatus{}, nil
	}

	return nil, *result, nil
}

// GoogleDriveConnect starts Google Drive OAuth flow.
func GoogleDriveConnect(ctx context.Context, client *labradoc.Client) (*mcp.CallToolResult, GoogleDriveConnectResult, error) {
	result, err := client.GoogleDriveConnect(ctx)
	if err != nil {
		return errorResult(err), labradoc.BillingCheckoutResponse{}, nil
	}

	return nil, *result, nil
}

// GoogleGmailStatus checks if Gmail is connected.
func GoogleGmailStatus(ctx context.Context, client *labradoc.Client) (*mcp.CallToolResult, GoogleGmailStatusResult, error) {
	result, err := client.GoogleGmailStatus(ctx)
	if err != nil {
		return errorResult(err), labradoc.IntegrationStatus{}, nil
	}

	return nil, *result, nil
}

// GoogleGmailConnect starts Gmail OAuth flow.
func GoogleGmailConnect(ctx context.Context, client *labradoc.Client) (*mcp.CallToolResult, GoogleGmailConnectResult, error) {
	result, err := client.GoogleGmailConnect(ctx)
	if err != nil {
		return errorResult(err), labradoc.BillingCheckoutResponse{}, nil
	}

	return nil, *result, nil
}

// MicrosoftOutlookStatus checks if Microsoft Outlook is connected.
func MicrosoftOutlookStatus(ctx context.Context, client *labradoc.Client) (*mcp.CallToolResult, MicrosoftOutlookStatusResult, error) {
	result, err := client.MicrosoftOutlookStatus(ctx)
	if err != nil {
		return errorResult(err), labradoc.IntegrationStatus{}, nil
	}

	return nil, *result, nil
}

// MicrosoftOutlookConnect starts Microsoft Outlook OAuth flow.
func MicrosoftOutlookConnect(ctx context.Context, client *labradoc.Client) (*mcp.CallToolResult, MicrosoftOutlookConnectResult, error) {
	result, err := client.MicrosoftOutlookConnect(ctx)
	if err != nil {
		return errorResult(err), labradoc.BillingCheckoutResponse{}, nil
	}

	return nil, *result, nil
}
